### PR TITLE
STM32 die temperature sensor driver rework

### PR DESCRIPTION
**Pull request by mathieuchopstm**
This PR is a rework of the STM32 die temperature sensor driver.

The main reason for this PR is the upcoming introduction of STM32WB0 series, which has a few differences from other series (calibration data is 32-bit, formula is different, sensor is always enabled) - with this reworked driver, the main differences are factored out, so the series can be easily integrated.

While at it, the driver and bindings have been heavily documented, and modified to follow the Reference Manuals more closely.

This PR also fixes errors in certain SoCs' DTSI (STM32030/STM32070 and STM32100 lines, STM32F2 series) and a few MISRA Rule 20.9 violations (using `#if xxx` on potentially undefined macros, instead of `#if defined(xxx)`).

### CODE REVIEW

File: drivers/sensor/st/stm32_temp/stm32_temp.c [hunk 1]
```diff
@@ -121,42 +121,93 @@ static void read_calibration_data(const struct stm32_temp_config *cfg,
...
+	temperature /= cfg->average_slope;
+	temperature += 25.0f;
+#else /* HAS_CALIBRATION */
 	uint32_t calib[MAX_CALIB_POINTS];
 
 	read_calibration_data(cfg, calib);
 
-	temperature = ((float)data->raw * adc_ref_internal(data->adc)) / cfg->calib_vrefanalog;
-	temperature -= (calib[0] >> cfg->calib_data_shift);
+	const float Sense_Data = ((float)vdda_mv / cfg->calib_vrefanalog) * data->raw;
```
Comment (erwango) [inline drivers/sensor/st/stm32_temp/stm32_temp.c:162]: Please stick with strict snake_case
```suggestion
	const float sense_data = ((float)vdda_mv / cfg->calib_vrefanalog) * data->raw;
```
Apply everywhere

File: dts/bindings/sensor/st,stm32-temp.yaml [hunk 2]
```diff
@@ -39,7 +39,7 @@ properties:
 
   v25:
     type: int
-    default: 760
+    required: true
```
Comment (erwango) [inline dts/bindings/sensor/st,stm32-temp.yaml:42]: Please move this in a different commit, along with potential impacts reflected in dtsi or a comment explaining that there is strangely no impact.

**Review by erwango — CHANGES_REQUESTED** [review]
Great effort, but some commit clean up required.

**Comment by mathieuchopstm** [comment]
CI failure is due to mistake in patch (`dietemp` for STM32F100 lines is not placed properly), will fix in V2.

**Comment by mathieuchopstm** [comment]
CI failures unrelated to PR content (twister failure caused by Renesas board, Compliance Checks seems to just be broken...)

**Comment by mathieuchopstm** [comment]
Commits 1~3 are minor bugfixes that I think are eligible for backport to 3.7 - should they be dropped from this PR and contributed by a separate PR for this purpose, if applicable?

* [`dts: st: stm32f2: remove 'ntc' property from die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/f0444ed8234eae7d4c2a904d088867db3aa0afb3)
* [`dts: st: stm32f030: add correct 'avgslope' property on die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/d4f4ab5b2341d65bee30b121ec5d9e0754f7971c)
* [`dts: st: stm32f100: add correct 'v25' property on die temp sensor`](https://github.com/zephyrproject-rtos/zephyr/pull/77610/commits/6c39d09da1dbad8339ddb42303df340dd642cffc)

**Review by erwango — APPROVED** [review]

**Review by gautierg-st — APPROVED** [review]

**Comment by erwango** [comment]
@MaureenHelm Kind reminder

**Review by MaureenHelm — APPROVED** [review]
